### PR TITLE
Ignore intent/entity prediction if input is empty

### LINKED ISSUE #10504: Empty (or some spaces) input text causes Runtime error
### Rasa Open Source version

3.0.0

### Rasa SDK version

3.0.0

### Rasa X version

_No response_

### Python version

3.8

### What operating system are you using?

Linux

### What happened?

In NLU shell, when I input text="" or text="     "(some spaces), it gets error as follow (full details in 'Relevant log output') :
`
rasa.engine.exceptions.GraphComponentException: Error running graph component for node run_DIETClassifier4.
`

But in RASA-2.8.6, everything goes well and gets following response:
`
{
  "text": "",
  "intent": {
    "name": null,
    "confidence": 0.0
  },
  "entities": []
}
`

I suppose this is a new bug.

### Command / Request

_No response_

### Relevant log output

```shell
Traceback (most recent call last):
File "~/venv/lib/python3.8/site-packages/rasa/engine/graph.py", line 461, in __call__
output = self._fn(self._component, **run_kwargs)
File "~/venv/lib/python3.8/site-packages/rasa/nlu/classifiers/diet_classifier.py", line 1003, in process
out = self._predict(message)
File "~/venv/lib/python3.8/site-packages/rasa/nlu/classifiers/diet_classifier.py", line 932, in _predict
return self.model.run_inference(model_data)
File "~/venv/lib/python3.8/site-packages/rasa/utils/tensorflow/models.py", line 318, in run_inference
] = self._rasa_predict(batch_in)
File "~/venv/lib/python3.8/site-packages/rasa/utils/tensorflow/models.py", line 280, in _rasa_predict
outputs = tf_utils.sync_to_numpy_or_python_type(self.predict_step(batch_in))
File "~/venv/lib/python3.8/site-packages/rasa/utils/tensorflow/models.py", line 228, in predict_step
return self.batch_predict(batch_in)
File "~/venv/lib/python3.8/site-packages/rasa/nlu/classifiers/diet_classifier.py", line 1721, in batch_predict
tf_batch_data = self.batch_to_model_data_format(
File "~/venv/lib/python3.8/site-packages/rasa/utils/tensorflow/models.py", line 473, in batch_to_model_data_format
tensor, idx = RasaModel._convert_sparse_features(
File "~/venv/lib/python3.8/site-packages/rasa/utils/tensorflow/models.py", line 518, in _convert_sparse_features
shape = [batch[idx + 2][i] for i in range(number_of_dimensions - 1)] + [
File "~/venv/lib/python3.8/site-packages/rasa/utils/tensorflow/models.py", line 518, in <listcomp>
shape = [batch[idx + 2][i] for i in range(number_of_dimensions - 1)] + [
IndexError: index 1 is out of bounds for axis 0 with size 1

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
File "~/venv/bin/rasa", line 8, in <module>
sys.exit(main())
File "~/venv/lib/python3.8/site-packages/rasa/__main__.py", line 121, in main
cmdline_arguments.func(cmdline_arguments)
File "~/venv/lib/python3.8/site-packages/rasa/cli/shell.py", line 91, in shell_nlu
rasa.nlu.run.run_cmdline(model)
File "~/venv/lib/python3.8/site-packages/rasa/nlu/run.py", line 25, in run_cmdline
result = asyncio.run(agent.parse_message(message))
File "/usr/lib/python3.8/asyncio/runners.py", line 44, in run
return loop.run_until_complete(main)
File "uvloop/loop.pyx", line 1456, in uvloop.loop.Loop.run_until_complete
File "~/venv/lib/python3.8/site-packages/rasa/core/agent.py", line 419, in parse_message
return await self.processor.parse_message(message)
File "~/venv/lib/python3.8/site-packages/rasa/core/processor.py", line 635, in parse_message
parse_data = self._parse_message_with_graph(message, only_output_properties)
File "~/venv/lib/python3.8/site-packages/rasa/core/processor.py", line 659, in _parse_message_with_graph
results = self.graph_runner.run(
File "~/venv/lib/python3.8/site-packages/rasa/engine/runner/dask.py", line 106, in run
dask_result = dask.get(run_graph, run_targets)
File "~/venv/lib/python3.8/site-packages/dask/local.py", line 560, in get_sync
return get_async(
File "~/venv/lib/python3.8/site-packages/dask/local.py", line 503, in get_async
for key, res_info, failed in queue_get(queue).result():
File "/usr/lib/python3.8/concurrent/futures/_base.py", line 437, in result
return self.__get_result()
File "/usr/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
raise self._exception
File "~/venv/lib/python3.8/site-packages/dask/local.py", line 545, in submit
fut.set_result(fn(*args, **kwargs))
File "~/venv/lib/python3.8/site-packages/dask/local.py", line 237, in batch_execute_tasks
return [execute_task(*a) for a in it]
File "~/venv/lib/python3.8/site-packages/dask/local.py", line 237, in <listcomp>
return [execute_task(*a) for a in it]
File "~/venv/lib/python3.8/site-packages/dask/local.py", line 228, in execute_task
result = pack_exception(e, dumps)
File "~/venv/lib/python3.8/site-packages/dask/local.py", line 223, in execute_task
result = _execute_task(task, data)
File "~/venv/lib/python3.8/site-packages/dask/core.py", line 121, in _execute_task
return func(*(_execute_task(a, cache) for a in args))
File "~/venv/lib/python3.8/site-packages/rasa/engine/graph.py", line 468, in __call__
raise GraphComponentException(
rasa.engine.exceptions.GraphComponentException: Error running graph component for node run_DIETClassifier4.
```

**Definition of done**
- [ ] Fix bug

### PR DESCRIPTION
**Pull request by jupyterjazz**
**Proposed changes**:
- Some components (e.g. `MitieFeaturizer`) can't handle empty strings which leads to [errors during intent prediction](https://github.com/RasaHQ/rasa/issues/10504). 
- Ignore intent/entity prediction in `DIETClassifier`, `CRFEntityExtractor`, `ResponseSelector` and `SklearnIntentClassifier` if we encounter an empty string.
- Fixes #10504 

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by dakshvar22** [comment]
What would happen inside core ensemble (for action prediction) if the input message is empty and hence the intent is missing?

**Comment by dakshvar22** [comment]
Also, if there was an original issue for which you opened the PR, could you please link it in the PR description?

**Comment by jupyterjazz** [comment]
> What would happen inside core ensemble (for action prediction) if the input message is empty and hence the intent is missing?

In that case, it's considered as `nlu_fallback`.

I also checked this with a default bot and looking at the `trackers_as_states`
https://github.com/RasaHQ/rasa/blob/e13bc1a77617cd557a9c8b92c4002f22bff72402/rasa/core/featurizers/tracker_featurizers.py#L425
which is used by policies for prediction, empty string gets represented the same way as it would be without this change.
`{'user': {'intent': 'nlu_fallback'}, 'prev_action': {'action_name': 'action_listen'}}`

**Comment by dakshvar22** [comment]
Okay, that makes sense. Another question: Looking at the original issue, it seems like the bug only comes up when you add `MitieFeaturizer` to the config. If the only change that was needed to fix the issue was in `DIETClassifier`, then why does the issue not come up in the config which does not have `MitieFeaturizer` in the config (for e.g. - `rasa init` config).

**Comment by jupyterjazz** [comment]
> then why does the issue not come up in the config which does not have `MitieFeaturizer` in the config

The reason is that for an empty string `MitieFeaturizer` doesn't create features which are expected during intent/entity prediction.

**Comment by dakshvar22** [comment]
Can you please also check if the fix that you added to DIET needs to be added to `CRFEntityExtractor` and `ResponseSelector`?

File: tests/nlu/classifiers/test_diet_classifier.py [hunk 1]
```diff
@@ -346,6 +346,17 @@ async def test_train_persist_load_with_only_intent_classification(
     create_diet({MASKED_LM: True, EPOCHS: 1}, load=True, finetune=True)
 
 
+async def test_process_empty_input(
+    create_train_load_and_process_diet: Callable[..., Message],
+):
+    message = create_train_load_and_process_diet(
+        diet_config={EPOCHS: 1}, message_text="", expect_intent=False
+    )
+    assert message.get(TEXT) == ""
```
Comment (dakshvar22) [inline tests/nlu/classifiers/test_diet_classifier.py:355]: Add a check for entities being empty as well

File: changelog/10504.bugfix.md [hunk 2]
```diff
@@ -0,0 +1 @@
+If NLU gets an empty string, ignore predicting intents and entities. Some components can't handle empty strings (e.g `MitieFeaturizer`) which causes errors during intent prediction.
```
Comment (dakshvar22) [inline changelog/10504.bugfix.md:1]:
```suggestion
Fixes a bug which caused machine learning components like `DIETClassifier`, `ResponseSelector`, `CRFEntityExtractor` and `SklearnIntentClassifier` to throw an error when a message with an empty text string was passed to them for processing.
```

**Comment by dakshvar22** [comment]
@jupyterjazz Almost there ✨  A couple of things:

1. I see that the original issue was tagged as a bug, so this should go out as a bugfix in a patch release. Hence, the PR should target `3.0.x` branch and should be forked from that branch. Currently its `main`.
2. Once you have done the changes, please run model regression tests overnight so that we are certain that inference is not broken.

**Pull request closed without merging**